libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1588,6 +1530,157 @@ internal class MapRouteLine(
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] = it.mapboxStreetsV8()?.roadClass()
+                                ?: "intersection_without_class_fallback"
+                        }
+
+                    var runningDistance = 0.0
+                    val routeLineTrafficData = mutableListOf<RouteLineTrafficExpressionData>()
+                    route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.forEachIndexed { index, congestion ->
+                        val roadclass = getRoadClassForIndex(roadClassMap, index)
+                        if (index == 0) {
+                            routeLineTrafficData.add(
+                                RouteLineTrafficExpressionData(
+                                    0.0,
+                                    congestion,
+                                    roadclass
+                                )
+                            )
+                        } else {
+                            runningDistance += distanceList[index - 1]
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                routeLineTrafficData.last().roadClass == roadclass
+                            ) {
+                                // continue
+                            } else if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                roadclass == null
+                            ) {
+                                // continue
+                            } else {
+                                routeLineTrafficData.add(
+                                    RouteLineTrafficExpressionData(
+                                        runningDistance,
+                                        congestion,
+                                        roadclass
+                                    )
+                                )
+                            }
+                        }
+                    }
+                    return routeLineTrafficData
+                }
+                Timber.i(
+                    "Route leg annotation does not contain distance and/or congestion " +
+                        "so route line traffic cannot be calculated."
+                )
+                return listOf()
+            }
+
+        private fun getRoadClassForIndex(roadClassMap: Map<Int, String>, index: Int): String? {
+            for (indexCountdown in index downTo 0) {
+                if (roadClassMap.containsKey(indexCountdown)) {
+                    return roadClassMap[indexCountdown]
+                }
+            }
+            return null
+        }
```

Comment:
This method seems to be very inefficient and takes up more than 90% of the main thread's time:
![Screenshot from 2020-11-05 13-57-30](https://user-images.githubusercontent.com/16925074/98243945-f3ad0c00-1f6e-11eb-91dd-ec3a364fe794.png)

I'm currently running into ANRs when trying to get a route from DC to LA.

I'll try to come up with some optimizations.

Reply:
Resolved with https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/db56e146905077f252201c4183a56b5a789a8483. The route still freezes the UI for ~2 seconds but it matches the experience on `master`. We should prioritize moving the operations to the worker thread with 1.3.

Reply:
Follow up in https://github.com/mapbox/mapbox-navigation-android/issues/3737.